libp2p-swarm/src/transport.rs

```diff
@@ -879,30 +809,22 @@ where
 		// Instead the `stream` will produce `Ok(Err(...))`.
 		// `stream` can only produce an `Err` if `listening_stream` produces an `Err`.
 		let stream = listening_stream
-			// Try to negotiate the protocol
-			.and_then(move |(connection, client_addr)| {
-				// Turn the `Result<impl AsyncRead + AsyncWrite, IoError>` into
-				// a `Result<impl Future<Item = impl AsyncRead + AsyncWrite, Error = IoError>, IoError>`
-				let connection = connection.map(|connection| {
-					let upgrade = upgrade.clone();
-					let iter = upgrade.protocol_names()
-						.map::<_, fn(_) -> _>(|(n, t)| (n, <Bytes as PartialEq>::eq, t));
-					multistream_select::listener_select_proto(connection, iter)
-						.map_err(|err| IoError::new(IoErrorKind::Other, err))
-						.and_then(|(upgrade_id, connection)| {
-							upgrade.upgrade(connection, upgrade_id, Endpoint::Listener)
-						})
-				});
-
-				connection
-					.into_future()
-					.flatten()
-					.then(move |nego_res| {
-						match nego_res {
-							Ok(upgraded) => Ok((Ok(upgraded), client_addr)),
-							Err(err) => Ok((Err(err), client_addr)),
-						}
-					})
+			.map(move |(connection, client_addr)| {
```

Comment:
Same as above, we can do this without `Box`

Reply:
We also move stuff in the closure here :-/